wafHook: always enable parallel building

panicgh opened the pull request:
###### Motivation for this change

Fixes #136368.

Basically, `wafHook` uses all available cores for building when `enableParallelBuilding=false` or unset and ignores `NIX_BUILD_CORES` (or `nix-build --cores <N>`).

The suggestion in https://github.com/NixOS/nixpkgs/issues/136368#issuecomment-910175545 would have set all builds sequential unless `enableParallelBuilding=true`. This would have caused follow-up changes to several pkgs as indicated in the issue. Furthermore, the [recommendation](https://nixos.org/manual/nixpkgs/stable/#attributes-affecting-build-properties) seems to prefer parallel building if the build system has good support for it.

This PR instead mimics the implementations of e.g. `cmake` or `meson` and sets `enablesParallelBuilding=1` if it has not been set otherwise. Dependent packages do not have to be modified, but would correctly honor `NIX_BUILD_CORES`.

The disadvantage of this proposal is that the build uses all available cores if `enableParallelBuilding` is explicitly set to `false`.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

veprbl commented:
This is same as in cmake:
https://github.com/NixOS/nixpkgs/blob/a8f9539d0171a5ae9285692805c0b206cfee9fde/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L124-L127
so must be good.

One thing: this needs to be rebased to staging, and, if possible, I would add a mention to the release notes.

panicgh committed:
wafHook: always enable parallel building

Follow the same pattern as used in the cmake setup hook to assume
`enableParallelBuilding=1` when unset by the user. This makes wafHook
honor `NIX_BUILD_CORES`. See #136368 and #136641.

panicgh commented:
Sorry for the long delay, I didn't find the time to follow this up earlier.

> One thing: this needs to be rebased to staging, and, if possible, I would add a mention to the release notes.

I rebased the branch onto staging and added a few words to the 21.11 release notes. Sorry for the noise due to the rebase. Hopefully this is better now.

SuperSandro2000 commented on pkgs/development/tools/build-managers/wafHook/setup-hook.sh:
```diff
@@ -21,6 +21,11 @@ wafConfigurePhase() {
 
+    if ! [[ -v enableParallelBuilding ]]; then
+        enableParallelBuilding=1
+        echo "waf: enabled parallel building"
```
```suggestion
```
don't think we need to log that tbh.

veprbl replied:
cmake does it, so it's fine to have here as well

panicgh replied:
I mark this conversation as resolved. Please reopen/open a new one if you strongly disagree and think further discussion is needed.

veprbl committed:
Partially revert "wafHook: always enable parallel building"

As a breaking change, this is unlikely to make it into 21.11. Let's
add this back again once 22.05 section is started.

This partially reverts commit 556ee949913f05f61290b9583edd76464dd3b0b5.

veprbl commented:
Sorry for delay. I don't think there is a need to port it into 21.11, let's merge to staging now. I've temporarily reverted your nice release notes additions, let's revisit this to add them back for 22.05 at a later time.

panicgh commented:
> I've temporarily reverted your nice release notes additions, let's revisit this to add them back for 22.05 at a later time.

Fine with me. Thanks for merging!
